File: scripts/check_dir_file_count.py
#!/usr/bin/env python3
"""
Check that no directory contains more than N files (default 10). Ignores `node_modules`.
Exit code 0 on success, 1 on failure.

Usage:
  python scripts/check_dir_file_count.py [root] [--max N]
"""
import os
import sys
import argparse


def parse_args():
    p = argparse.ArgumentParser(description="Check directories for file count")
    p.add_argument("root", nargs="?", default=".", help="Root directory to scan")
    p.add_argument(
        "--max", "-m", type=int, default=10, help="Maximum allowed files per directory"
    )
    p.add_argument(
        "--ignore",
        "-i",
        action="append",
        default=["node_modules", ".git", "dist"],
        help="Directory names to ignore (repeatable)",
    )
    return p.parse_args()


def main():
    args = parse_args()
    offenders = []

    for dirpath, dirnames, filenames in os.walk(args.root):
        # skip ignored directories immediately
        dirnames[:] = [d for d in dirnames if d not in args.ignore]
        # Filter out compiled TypeScript files (.js and .d.ts) to count only source files
        source_files = [
            f for f in filenames if not f.endswith(".js") and not f.endswith(".d.ts")
        ]
        file_count = len(source_files)
        if file_count > args.max:
            offenders.append((dirpath, file_count, source_files))

    if not offenders:
        print(f"OK: no directories with more than {args.max} files under '{args.root}'")
        return 0

    print(f"Found {len(offenders)} directories exceeding {args.max} files:")
    for path, count, files in offenders:
        print(f"- {path}: {count} files")
        sample = files[:20]
        for f in sample:
            print(f"    {f}")
        if len(files) > len(sample):
            print(f"    ... and {len(files)-len(sample)} more files")

    return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-dir-file-count.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");

const MAX_FILES_PER_DIR = 10;
const EXCLUDE_DIRS = new Set(["node_modules", ".git", "dist", "build", ".husky"]);

function countFilesInDir(dirPath) {
  try {
    const entries = fs.readdirSync(dirPath);
    const files = entries.filter((entry) => {
      const fullPath = path.join(dirPath, entry);
      const stat = fs.statSync(fullPath);
      return stat.isFile();
    });
    return files.length;
  } catch (err) {
    console.error(`Error reading directory ${dirPath}:`, err.message);
    return 0;
  }
}

function checkDirectoriesRecursively(dirPath, relativePath = "") {
  const entries = fs.readdirSync(dirPath);
  let allPass = true;

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry);
    const currentRelPath = relativePath ? `${relativePath}/${entry}` : entry;

    // Skip excluded directories
    if (EXCLUDE_DIRS.has(entry)) {
      continue;
    }

    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      // Check file count in this directory
      const fileCount = countFilesInDir(fullPath);
      if (fileCount > MAX_FILES_PER_DIR) {
        console.error(
          `✗ Directory '${currentRelPath}' has ${fileCount} files (max: ${MAX_FILES_PER_DIR})`
        );
        allPass = false;
      } else if (fileCount > 0) {
        console.log(
          `✓ Directory '${currentRelPath}' has ${fileCount} files (OK)`
        );
      }

      // Recurse into subdirectories
      if (!checkDirectoriesRecursively(fullPath, currentRelPath)) {
        allPass = false;
      }
    }
  }

  return allPass;
}

console.log(
  `Checking directory file counts (max ${MAX_FILES_PER_DIR} files per directory)...\n`
);

const rootDir = process.cwd();
const allPass = checkDirectoriesRecursively(rootDir);

if (!allPass) {
  console.error("\n✗ Some directories exceed the file limit!");
  process.exit(1);
}

console.log("\n✓ All directories pass the file count check!");
process.exit(0);
